### ISSUE TITLE
`import HTTPTypesFoundation` pulls in `FoundationNetworking`

### ISSUE THREAD

**Issue by weissi** [apple/swift-http-types#76]
Unfortunately, `import HTTPTypesFoundation` pulls in `FoundationNetworking` unconditionally which has [issues on Linux](https://github.com/swiftlang/swift-corelibs-foundation/issues?q=is%3Aissue+is%3Aopen+URLSession).

We carry link-time checks to make sure that we're not accidentally using `URLSession` in places where it's important not to. Unfortunately, modules that `import HTTPTypesFoundation` now also pull in `FoundationNetworking`.

Would it be possible to remove the `import FoundationNetworking` and move that to a separate module like `HTTPTypesFoundationNetworking`?

**Comment by guoye-zhang** [comment]
Does this require a major version bump due to source breakages?

**Comment by Lukasa** [comment]
Yeah, unfortunately it does due to the leaky nature of Swift imports. In this case it a "cheap" major, and we can encourage users to depend on a cross-major range.

**Comment by weissi** [comment]
> Yeah, unfortunately it does due to the leaky nature of Swift imports. In this case it a "cheap" major, and we can encourage users to depend on a cross-major range.

Yes, a "cheap major" is perfect here. We've done this before and asking users to do `"1.0.0" ..< "3.0.0"` is usually not a problem. Especially because it can be done at any point in time and doesn't break stuff either way.

**Comment by guoye-zhang** [comment]
I checked and URLRequest and URLResponse are both in FoundationNetworking. `HTTPTypesFoundation` would have no API at all if it doesn't import FoundationNetworking.

Why do these packages `import HTTPTypesFoundation`?

**Comment by weissi** [comment]
> I checked and URLRequest and URLResponse are both in FoundationNetworking. `HTTPTypesFoundation` would have no API at all if it doesn't import FoundationNetworking.
> 
> Why do these packages `import HTTPTypesFoundation`?

Just for the `extension HTTPRequest { public var url: URL? }` :)

**Comment by guoye-zhang** [comment]
Oh I see

**Comment by parkera** [comment]
`URL` is part of `FoundationEssentials`, so that could be the only requirement there.

**Comment by guoye-zhang** [comment]
Alternatively, can we move `extension HTTPRequest { public var url: URL? }` to the main HTTPTypes package? Can we assume that `FoundationEssentials` exists everywhere?

**Comment by weissi** [comment]
> Alternatively, can we move `extension HTTPRequest { public var url: URL? }` to the main HTTPTypes package? Can we assume that `FoundationEssentials` exists everywhere?

I don't think `NIOExtras` has `FoundationEssentials`, @Lukasa?

**Comment by Lukasa** [comment]
Correct, currently we don't use `FoundationEssentials` in NIO-world. I'm not sure how long that state of affairs will last though: the existence of `FoundationEssentials` relieves many of the constraints of depending on `Foundation`.

**Comment by parkera** [comment]
FoundationEssentials is present on all non-Darwin platforms as part of the Swift toolchain. On Darwin it's still just Foundation.

**Comment by guoye-zhang** [comment]
What's the difference between `import Foundation` and `import FoundationEssentials` on non-Darwin? Does `import FoundationEssentials` work on Swift 5.7 or is it more recent?

**Comment by parkera** [comment]
`import FoundationEssentials` is available in Swift 6.0 or later. See [here](https://github.com/swiftlang/swift-foundation?tab=readme-ov-file#swift-foundation) for more information on how these are all layered.

**Comment by guoye-zhang** [comment]
Would you recommend that we do something like:

```
#if canImport(FoundationEssentials)
import FoundationEssentials
#else
import Foundation
#endif
```

**Comment by FranzBusch** [comment]
Yes that's the recommend pattern to check if it exists and import only that. IMO that's the way we should do it here and move the extension for URL to the main module of this package behind the conditional import.

**Comment by guoye-zhang** [comment]
Are people all OK with moving URL to the main package? I can make a new PR to do that
